File: spec/models/device_spec.rb
# spec/models/device_spec.rb
# Copyright 2014 Sergei Ianovich
#
# Licensed under AGPLv3 or later
# Process Control Service Web Interface

require 'spec_helper'

describe Device do
  before do
    Mongoid.default_session.collections.select {|c| c.name !~ /system/}.each {|c| c.find.remove_all}
    @device = Device.new(name: "a device", filepath: "/dev/null")
  end

  subject { @device }

  it { should respond_to(:name) }
  it { should respond_to(:filepath) }
  it { should respond_to(:enabled) }
  it { should respond_to(:read_new_states) }

  it { should be_valid }

  describe "when name is not present" do
    before { @device.name = " " }
    it { should_not be_valid }
  end

  describe "when filepath is not present" do
    before { @device.filepath = " " }
    it { should_not be_valid }
  end
end

# vim:ts=2 sts=2 sw=2 et:

File: test/backend/models/device_spec.js
'use strict'
/* test/backend/models/device.js -- test Device model
 * Copyright 2014 Sergei Ianovich
 *
 * Licensed under AGPL-3.0 or later, see LICENSE
 * Process Control Service Web Interface
 */

var helper = require('../spec_helper');
var Device = require('../../../models/device');

var deviceAttrs = { name: "Example Device" }

describe('Device', function () {
  var device;
  beforeEach(function (done) {
    device = new Device(deviceAttrs);
    Device.find().remove(done);
  })

  it('should respond to name', function () {
    expect(device.name).toBeDefined();
  });

  it('should be valid', function (done) {
    device.validate(function (err) {
      expect(err).toBeFalsy();
      done();
    });
  });
  
  describe('when name is not present', function () {
    it('should not be valid', function (done) {
      device.name = " ";
      device.validate(function(err) {
        expect(err).toBeTruthy();
        done();
      });
    });
  })

  describe('when name is too long', function () {
    it('should not be valid', function (done) {
      device.name = "";
      for (var i = 0; i < 51; ++i) {
        device.name += "a";
      }
      device.validate(function(err) {
        expect(err).toBeTruthy();
        done();
      });
    });
  })

  describe('when name already exists', function () {
    it('should not save', function (done) {
      device.save(function(err, dev) {
        expect(err).toBeFalsy();
        device = new Device(deviceAttrs);
        device.save(function(err) {
          expect(err).toBeTruthy();;
          done();
        });
      });
      
    });
  })

});

